rust-runtime/aws-smithy-http-server/src/operation/mod.rs

```diff
@@ -0,0 +1,268 @@
+//! argument, the remaining arguments implement [`FromParts`](crate::request::FromParts), and return either
+//! [`OperationShape::Output`] when [`OperationShape::Error`] is [`Infallible`](std::convert::Infallible) or
+//! [`Result`]<[`OperationShape::Output`],[`OperationShape::Error`]>. The following are examples of closures which
+//! implement [`Handler`]:
+//!
+//! ```rust
+//! # use aws_smithy_http_server::Extension;
+//! # pub struct CartIdentifier;
+//! # pub struct ShoppingCart;
+//! # pub enum GetShoppingError {}
+//! # pub struct Context;
+//! # pub struct ExtraContext;
+//! // Simple handler where no error is given.
+//! async fn handler_a(input: CartIdentifier) -> ShoppingCart {
+//!     todo!()
+//! }
+//!
+//! // Handler with an extension where no error is given.
+//! async fn handler_b(input: CartIdentifier, ext: Extension<Context>) -> ShoppingCart {
+//!     todo!()
+//! }
+//!
+//! // More than one extension can be provided.
+//! async fn handler_c(input: CartIdentifier, ext_1: Extension<Context>, ext_2: Extension<ExtraContext>) -> ShoppingCart {
+//!     todo!()
+//! }
+//!
+//! // When an error is given we must return a `Result`.
+//! async fn handler_d(input: CartIdentifier, ext: Extension<Context>) -> Result<ShoppingCart, GetShoppingError> {
+//!     todo!()
+//! }
+//! ```
+//!
+//! ## [`OperationService`]
+//!
+//! Similarly, the [`OperationService`] trait is implemented by all `Service<(Op::Input, ...)>` with
+//! `Response = Op::Output`, and `Error = OperationError<Op::Output, PollError>`.
+//!
+//! We use [`OperationError`], with a `PollError` not constrained by the model, to allow the user to provide a custom
+//! [`Service::poll_ready`](tower::Service::poll_ready) implementation.
+//!
+//! The following are examples of [`Service`](tower::Service)s which implement [`OperationService`]:
+//!
+//! - `Service<CartIdentifier, Response = ShoppingCart, Error = Infallible>`.
+//! - `Service<(CartIdentifier, Extension<Context>), Response = ShoppingCart, Error = GetShoppingCartError>`.
+//! - `Service<(CartIdentifier, Extension<Context>, Extension<ExtraContext>), Response = ShoppingCart, Error = GetShoppingCartError)`.
+//!
+//! Notice the parallels between [`OperationService`] and [`Handler`].
+//!
+//! ## Constructing an [`Operation`]
+//!
+//! The following is an example of using both construction approaches:
+//!
+//! ```rust
+//! # use std::task::{Poll, Context};
+//! # use aws_smithy_http_server::operation::*;
+//! # use tower::Service;
+//! # pub struct CartIdentifier;
+//! # pub struct ShoppingCart;
+//! # pub enum GetShoppingError {}
+//! # pub struct GetShopping;
+//! # impl OperationShape for GetShopping {
+//! #    const NAME: &'static str = "GetShopping";
+//! #
+//! #    type Input = CartIdentifier;
+//! #    type Output = ShoppingCart;
+//! #    type Error = GetShoppingError;
+//! # }
+//! # type OpFuture = std::future::Ready<Result<ShoppingCart, OperationError<GetShoppingError, PollError>>>;
```

Comment:
Isn't `OpFuture` something users have to define themselves? Shouldn't this appear in the example then?

Reply:
It is, but given that we haven't bothered to actually write a meaningful impl and this is just introduced to get it to type check I decided to keep it hidden.